test/k8sT/label_changes.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sTest
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	. "github.com/cilium/cilium/test/ginkgo-ext"
+	"github.com/cilium/cilium/test/helpers"
+
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("K8sPolicyTest", func() {
+
+	var (
+		kubectl          *helpers.Kubectl
+		demoPath                            = helpers.ManifestGet("demo.yaml")
+		backgroundCancel context.CancelFunc = func() { return }
+		backgroundError  error
+	)
+
+	BeforeAll(func() {
+		kubectl = helpers.CreateKubectl(helpers.K8s1VMName(), logger)
+
+		err := kubectl.CiliumInstall(helpers.CiliumDSPath)
+		Expect(err).To(BeNil(), "Cilium cannot be installed")
+
+		ExpectCiliumReady(kubectl)
+		ExpectKubeDNSReady(kubectl)
+	})
+
+	AfterEach(func() {
+		ExpectAllPodsTerminated(kubectl)
+	})
+
+	AfterFailed(func() {
+		kubectl.CiliumReport(helpers.KubeSystemNamespace,
+			"cilium service list",
+			"cilium endpoint list")
+	})
+
+	AfterAll(func() {
+		_ = kubectl.WaitCleanAllTerminatingPods()
+		_ = kubectl.Exec(fmt.Sprintf("%s label ns/%s foo-bar-",
+			helpers.KubectlCmd, helpers.DefaultNamespace))
+	})
+
+	JustBeforeEach(func() {
+		backgroundCancel, backgroundError = kubectl.BackgroundReport("uptime")
+		Expect(backgroundError).To(BeNil(), "Cannot start background report process")
+	})
+
+	JustAfterEach(func() {
+		kubectl.ValidateNoErrorsOnLogs(CurrentGinkgoTestDescription().Duration)
+		backgroundCancel()
+	})
+
+	Context("Cilium Endpoint Label Changes", func() {
+		BeforeAll(func() {
+			kubectl.Apply(demoPath).ExpectSuccess("Unable to create demo pods")
+			err := kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", 300)
+			Expect(err).Should(BeNil(), "Test pods are not ready after timeout")
+			err = kubectl.CiliumEndpointWaitReady()
+			Expect(err).Should(BeNil(), "Cilium Endpoints are not ready after timeout")
+		})
+
+		AfterAll(func() {
+			kubectl.Delete(demoPath)
+		})
+
+		It("Changes Endpoint Labels on Pod and Namespace Label Changes", func() {
+			checkIfPodsHaveLabels := func(podAnnotationPrev map[string]string) map[string]string {
```

Comment:
If you change this function to something like: `CheckIfPodLabelsUpdated(podAnnotationPrev map[string]string) newLabels map[string]string, updated bool)` you could then use it to get the first `podAnnotationPrev` and avoid the code duplication.
You would also move the `Expect` to the test. This seems better to me but I can see why you want it where it is.